source/contribute.rst

```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
+- ``2.4.x`` from which the next minor release will be tagged (2.4.0);
+- ``3.0.x`` from which the next major release will be tagged (3.0.0).
 
-Keeping the default branch up-to-date!
---------------------------------------
+Anything you contribute to a branch will be merged up to higher branches
+at some point. For example, if you contribute to ``2.3.x``, your changes
+will eventually land on ``2.4.x`` and ``3.0.x``.
 
-Once all this is done, you'll be able to keep your local branches up to
-date with the following command:
+The constraints we have are:
 
-.. code-block:: console
+- Every patch release should be more stable than the previous one.
+- An upgrade to a new minor version should not require any change on the
+  user side.
+- There should be a clear upgrade path from one major version to the
+  next.
+- We want to keep the branches as similar as possible to ease
+  maintenance.
 
-    $ git fetch doctrine
+With that in mind:
 
-Branching Model
----------------
+If you are touching something outside the source directory (usually
+``src``, sometimes ``lib``), and that thing is not ``composer.json``,
+you should probably target the lowest branch.
+There is no way that your contribution is going to create a bug for
+the end user, so let us get it on all 3 branches with merges up.
+Examples are: documentation, tests, build scripts, tooling configuration.
 
-The following names will be used to differentiate between the different
-repositories:
+If you are touching something inside the source directory, you may
+still contribute to the lowest branch if it is a bug fix, and if you
+are touching comments that are not phpdoc comments.
 
--  **doctrine** - The "official" Doctrine DBAL repository
--  **origin** - Your fork of the official repository on GitHub
--  **local** - This will be your local clone of **origin**
+If the comments are phpdoc comments… it depends. If the phpdoc was
+inaccurate, you should target the lowest branch. If the phpdoc was
+imprecise but technically correct, you should target the next minor
+branch.
 
-As a **contributor** you will push your completed **local** topic branch
-to **origin**. As a **contributor** you will pull updates from
-**doctrine**. As a **maintainer** (write-access) you will merge branches
-from contributors into **doctrine**.
+If you are doing a refactoring and you think you didn't change the
+behavior of the code, you should still target the next minor branch
+unless the refactoring is necessary for a bugfix. Let us not take any
+unnecessary risks.
 
-Primary Branches
-----------------
+If you want to deprecate something, you should target the next minor
+branch: upgrading to a new patch version should not result in any
+additional deprecations.
 
-The **doctrine** repository holds the following primary branches:
+If you want to add a new feature, you should target the next minor
+branch, because it might have side effects that break existing code.
```

Comment:
```suggestion
If you want to add a new feature, you should target the next minor
branch, because it might have unintentional side effects that break existing code.
```